plugins/techdocs-backend/src/search/DefaultTechDocsCollator.ts

```diff
@@ -0,0 +1,123 @@
+import fetch from 'cross-fetch';
+import unescape from 'lodash/unescape';
+import { Logger } from 'winston';
+
+export interface TechDocsDocument extends IndexableDocument {
+  kind: string;
+  namespace: string;
+  entityKind: string;
+  entity: string;
+  lifecycle: string;
+}
+
+export class DefaultTechDocsCollator implements DocumentCollator {
+  protected discovery: PluginEndpointDiscovery;
+  protected locationTemplate: string;
+  private readonly logger: Logger;
+  public readonly type: string = 'tech-docs';
+
+  constructor({
+    discovery,
+    locationTemplate,
+    logger,
+  }: {
+    discovery: PluginEndpointDiscovery;
+    locationTemplate?: string;
+    logger: Logger;
+  }) {
+    this.discovery = discovery;
+    this.locationTemplate =
+      locationTemplate || '/docs/:namespace/:kind/:name/:path';
+    this.logger = logger;
+  }
+
+  async execute() {
+    const catalogBaseUrl = await this.discovery.getBaseUrl('catalog');
+    const techDocsBaseUrl = await this.discovery.getBaseUrl('techdocs');
+    const res = await fetch(`${catalogBaseUrl}/entities`);
+    const entities: Entity[] = await res.json();
+    const docPromises = entities
+      .filter(it => it.metadata?.annotations?.['backstage.io/techdocs-ref'])
+      .map(
+        async (entity: Entity): Promise<TechDocsDocument[]> => {
+          const entityInfo = {
+            kind: entity.kind,
+            namespace: entity.metadata.namespace || 'default',
+            name: entity.metadata.name,
+          };
+
+          try {
+            const searchIndexResponse = await fetch(
+              DefaultTechDocsCollator.constructDocsIndexUrl(
+                techDocsBaseUrl,
+                entityInfo,
+              ),
+            );
+            const searchIndex = await searchIndexResponse.json();
+
+            return searchIndex.docs.map((doc: TechDocsDocument) => ({
+              title: unescape(doc.title),
+              text: unescape(doc.text || ''),
+              location: this.applyArgsToFormat(this.locationTemplate, {
+                namespace: entity.metadata.namespace || 'default',
+                kind: entity.kind,
+                name: entity.metadata.name,
+                path: doc.location,
+              }),
+              kind: 'Documentation',
+
+              componentType: entity.spec?.type?.toString() || 'other',
+              namespace: entity.metadata.namespace || 'default',
+              entityKind: entity.kind,
+              lifecycle: (entity.spec?.lifecycle as string) || '',
+              owner: (entity.spec?.owner as string) || '',
+            }));
+          } catch (e) {
+            this.logger.warn(
+              `Failed to retrieve tech docs search index for entity ${entityInfo.namespace}/${entityInfo.kind}/${entityInfo.name}`,
+              e,
+            );
+            return [];
+          }
+        },
+      );
+    return (await Promise.all(docPromises)).flat();
```

Comment:
Would advice to use p-limit here, otherwise you might do a lot of work in parallel

Reply:
Good shout, I'll add some limit in :+1: